File: apps/api/src/modules/product-analyze/domain/product-facts/product-type-utils.ts
import type { NormalizedProduct, ProductType } from '@acme/shared';

interface ProductTypeRule {
  type: ProductType;
  keywords: string[];
}

const PRODUCT_TYPE_RULES: ProductTypeRule[] = [
  {
    type: 'beverage',
    keywords: [
      'beverage',
      'drink',
      'juice',
      'soda',
      'water',
      'tea',
      'coffee',
      'milk drink',
      'smoothie',
      'energy drink',
      'boisson',
      'getränk',
    ],
  },
  {
    type: 'yogurt',
    keywords: ['yogurt', 'yoghurt', 'yaourt', 'joghurt', 'skyr'],
  },
  {
    type: 'cheese',
    keywords: ['cheese', 'fromage', 'käse', 'queso', 'formaggio'],
  },
  {
    type: 'dairy',
    keywords: ['dairy', 'milk', 'cream', 'butter', 'lait', 'milch', 'crème'],
  },
  {
    type: 'meat',
    keywords: [
      'meat',
      'chicken',
      'beef',
      'pork',
      'turkey',
      'lamb',
      'sausage',
      'ham',
      'viande',
      'fleisch',
    ],
  },
  {
    type: 'fish',
    keywords: [
      'fish',
      'salmon',
      'tuna',
      'shrimp',
      'seafood',
      'poisson',
      'fisch',
    ],
  },
  {
    type: 'bread',
    keywords: ['bread', 'baguette', 'toast', 'pain', 'brot', 'roll', 'pita'],
  },
  {
    type: 'cereal',
    keywords: [
      'cereal',
      'muesli',
      'granola',
      'oat',
      'porridge',
      'flakes',
      'céréale',
    ],
  },
  {
    type: 'sauce',
    keywords: [
      'sauce',
      'ketchup',
      'mustard',
      'mayonnaise',
      'dressing',
      'vinaigrette',
      'soße',
    ],
  },
  {
    type: 'sweet',
    keywords: [
      'candy',
      'chocolate',
      'confectionery',
      'bonbon',
      'gummy',
      'süßigkeit',
      'praline',
    ],
  },
  {
    type: 'dessert',
    keywords: [
      'dessert',
      'pudding',
      'ice cream',
      'mousse',
      'cake',
      'pastry',
      'pie',
      'tart',
    ],
  },
  {
    type: 'snack',
    keywords: [
      'snack',
      'chip',
      'crisp',
      'cracker',
      'pretzel',
      'popcorn',
      'nut mix',
      'bar',
    ],
  },
  {
    type: 'ready_meal',
    keywords: [
      'ready meal',
      'prepared',
      'frozen meal',
      'microwave',
      'pizza',
      'lasagna',
      'plat préparé',
    ],
  },
  {
    type: 'plant_protein',
    keywords: [
      'tofu',
      'tempeh',
      'seitan',
      'plant-based',
      'vegan meat',
      'soy protein',
      'protéine végétale',
    ],
  },
  {
    type: 'fruit_vegetable',
    keywords: [
      'fruit',
      'vegetable',
      'salad',
      'légume',
      'obst',
      'gemüse',
      'compote',
    ],
  },
];

/**
 * Detect product type from normalized product categories and name.
 * Uses keyword matching against category tags, categories string, and product name.
 */
export const detectProductType = (
  product: NormalizedProduct,
): ProductType | null => {
  const searchTexts = [
    ...(product.category_tags ?? []),
    product.categories ?? '',
    product.product_name ?? '',
  ]
    .map((t) => t.toLowerCase())
    .filter(Boolean);

  if (searchTexts.length === 0) return null;

  const combined = searchTexts.join(' ');

  for (const rule of PRODUCT_TYPE_RULES) {
    for (const keyword of rule.keywords) {
      if (combined.includes(keyword)) {
        return rule.type;
      }
    }
  }

  return 'other';
};

File: apps/server-py/app/domain/product_facts/product_type_utils.py
from typing import Optional

from app.domain.product_facts.schema import NormalizedProduct, ProductType

_PRODUCT_TYPE_RULES: list[tuple[ProductType, list[str]]] = [
    (
        "beverage",
        [
            "beverage",
            "drink",
            "juice",
            "soda",
            "water",
            "tea",
            "coffee",
            "milk drink",
            "smoothie",
            "energy drink",
            "boisson",
            "getränk",
        ],
    ),
    ("yogurt", ["yogurt", "yoghurt", "yaourt", "joghurt", "skyr"]),
    ("cheese", ["cheese", "fromage", "käse", "queso", "formaggio"]),
    ("dairy", ["dairy", "milk", "cream", "butter", "lait", "milch", "crème"]),
    ("meat", ["meat", "chicken", "beef", "pork", "turkey", "lamb", "sausage", "ham", "viande", "fleisch"]),
    ("fish", ["fish", "salmon", "tuna", "shrimp", "seafood", "poisson", "fisch"]),
    ("bread", ["bread", "baguette", "toast", "pain", "brot", "roll", "pita"]),
    ("cereal", ["cereal", "muesli", "granola", "oat", "porridge", "flakes", "céréale"]),
    ("sauce", ["sauce", "ketchup", "mustard", "mayonnaise", "dressing", "vinaigrette", "soße"]),
    ("sweet", ["candy", "chocolate", "confectionery", "bonbon", "gummy", "süßigkeit", "praline"]),
    ("dessert", ["dessert", "pudding", "ice cream", "mousse", "cake", "pastry", "pie", "tart"]),
    ("snack", ["snack", "chip", "crisp", "cracker", "pretzel", "popcorn", "nut mix", "bar"]),
    ("ready_meal", ["ready meal", "prepared", "frozen meal", "microwave", "pizza", "lasagna", "plat préparé"]),
    ("plant_protein", ["tofu", "tempeh", "seitan", "plant-based", "vegan meat", "soy protein", "protéine végétale"]),
    ("fruit_vegetable", ["fruit", "vegetable", "salad", "légume", "obst", "gemüse", "compote"]),
]


def detect_product_type(product: NormalizedProduct) -> Optional[ProductType]:
    search_texts = [
        *[t.lower() for t in (product.get("category_tags") or [])],
        (product.get("categories") or "").lower(),
        (product.get("product_name") or "").lower(),
    ]
    search_texts = [t for t in search_texts if t]

    if not search_texts:
        return None

    combined = " ".join(search_texts)

    for product_type, keywords in _PRODUCT_TYPE_RULES:
        for keyword in keywords:
            if keyword in combined:
                return product_type

    return "other"
